### PR TITLE
Fix mypy type checking errors in test_transaction.py

### PR DESCRIPTION
**Pull request by jameshilliard**
This fixes a number of issues with tests using the wrong types that was flagged by mypy.

### CODE REVIEW

**Comment by jameshilliard** [comment]
Errors on `dev` branch this fixes for reference:
```python
$ mypy test/test_transaction.py 
test/test_transaction.py:25: error: Need type annotation for "client" (hint: "client: <type> | None = ...")  [var-annotated]
test/test_transaction.py:26: error: Need type annotation for "decoder" (hint: "decoder: <type> | None = ...")  [var-annotated]
test/test_transaction.py:27: error: Need type annotation for "_tcp" (hint: "_tcp: <type> | None = ...")  [var-annotated]
test/test_transaction.py:28: error: Need type annotation for "_tls" (hint: "_tls: <type> | None = ...")  [var-annotated]
test/test_transaction.py:29: error: Need type annotation for "_rtu" (hint: "_rtu: <type> | None = ...")  [var-annotated]
test/test_transaction.py:30: error: Need type annotation for "_ascii" (hint: "_ascii: <type> | None = ...")  [var-annotated]
test/test_transaction.py:31: error: Need type annotation for "_manager" (hint: "_manager: <type> | None = ...")  [var-annotated]
test/test_transaction.py:32: error: Need type annotation for "_tm" (hint: "_tm: <type> | None = ...")  [var-annotated]
test/test_transaction.py:45: error: Argument 1 to "SyncModbusTransactionManager" has incompatible type "None"; expected "ModbusBaseSyncClient"  [arg-type]
test/test_transaction.py:53: error: Item "None" of "Any | None" has no attribute "client"  [union-attr]
test/test_transaction.py:54: error: Item "None" of "Any | None" has no attribute "client"  [union-attr]
test/test_transaction.py:55: error: Item "None" of "Any | None" has no attribute "_set_adu_size"  [union-attr]
test/test_transaction.py:56: error: Item "None" of "Any | None" has no attribute "_calculate_response_length"  [union-attr]
test/test_transaction.py:59: error: Item "None" of "Any | None" has no attribute "base_adu_size"  [union-attr]
test/test_transaction.py:61: error: Item "None" of "Any | None" has no attribute "_calculate_response_length"  [union-attr]
test/test_transaction.py:74: error: Item "None" of "Any | None" has no attribute "client"  [union-attr]
test/test_transaction.py:76: error: Item "None" of "Any | None" has no attribute "client"  [union-attr]
test/test_transaction.py:78: error: Item "None" of "Any | None" has no attribute "client"  [union-attr]
test/test_transaction.py:80: error: Item "None" of "Any | None" has no attribute "client"  [union-attr]
test/test_transaction.py:82: error: Item "None" of "Any | None" has no attribute "client"  [union-attr]
test/test_transaction.py:84: error: Item "None" of "Any | None" has no attribute "client"  [union-attr]
test/test_transaction.py:86: error: Item "None" of "Any | None" has no attribute "_set_adu_size"  [union-attr]
test/test_transaction.py:88: error: Item "None" of "Any | None" has no attribute "_calculate_exception_length"  [union-attr]
test/test_transaction.py:99: error: Item "None" of "Any | None" has no attribute "_buffer"  [union-attr]
test/test_transaction.py:100: error: Item "None" of "Any | None" has no attribute "processIncomingPacket"  [union-attr]
test/test_transaction.py:101: error: Item "None" of "Any | None" has no attribute "processIncomingPacket"  [union-attr]
test/test_transaction.py:102: error: Item "None" of "Any | None" has no attribute "buildPacket"  [union-attr]
test/test_transaction.py:103: error: Item "None" of "Any | None" has no attribute "buildPacket"  [union-attr]
test/test_transaction.py:104: error: Item "None" of "Any | None" has no attribute "sendPacket"  [union-attr]
test/test_transaction.py:105: error: Item "None" of "Any | None" has no attribute "sendPacket"  [union-attr]
test/test_transaction.py:106: error: Item "None" of "Any | None" has no attribute "decode_data"  [union-attr]
test/test_transaction.py:107: error: Item "None" of "Any | None" has no attribute "decode_data"  [union-attr]
test/test_transaction.py:117: error: Cannot assign to a method  [method-assign]
test/test_transaction.py:123: error: Cannot assign to a method  [method-assign]
test/test_transaction.py:128: error: Cannot assign to a method  [method-assign]
test/test_transaction.py:132: error: Cannot assign to a method  [method-assign]
test/test_transaction.py:139: error: Cannot assign to a method  [method-assign]
test/test_transaction.py:146: error: Cannot assign to a method  [method-assign]
test/test_transaction.py:157: error: Cannot assign to a method  [method-assign]
test/test_transaction.py:164: error: Cannot assign to a method  [method-assign]
test/test_transaction.py:167: error: Item "None" of "Any | None" has no attribute "processIncomingPacket"  [union-attr]
test/test_transaction.py:182: error: Cannot assign to a method  [method-assign]
test/test_transaction.py:191: error: Item "None" of "Any | None" has no attribute "getNextTID"  [union-attr]
test/test_transaction.py:192: error: Item "None" of "Any | None" has no attribute "getNextTID"  [union-attr]
test/test_transaction.py:193: error: Item "None" of "Any | None" has no attribute "reset"  [union-attr]
test/test_transaction.py:194: error: Item "None" of "Any | None" has no attribute "getNextTID"  [union-attr]
test/test_transaction.py:202: error: Item "None" of "Any | None" has no attribute "reset"  [union-attr]
test/test_transaction.py:204: error: "Request" has no attribute "transaction_id"  [attr-defined]
test/test_transaction.py:205: error: Item "None" of "Any | None" has no attribute "getNextTID"  [union-attr]
test/test_transaction.py:207: error: "Request" has no attribute "message"  [attr-defined]
test/test_transaction.py:208: error: Item "None" of "Any | None" has no attribute "addTransaction"  [union-attr]
test/test_transaction.py:209: error: Item "None" of "Any | None" has no attribute "getTransaction"  [union-attr]
test/test_transaction.py:209: error: "Request" has no attribute "transaction_id"  [attr-defined]
test/test_transaction.py:210: error: "Request" has no attribute "message"  [attr-defined]
test/test_transaction.py:218: error: Item "None" of "Any | None" has no attribute "reset"  [union-attr]
test/test_transaction.py:220: error: "Request" has no attribute "transaction_id"  [attr-defined]
test/test_transaction.py:221: error: Item "None" of "Any | None" has no attribute "getNextTID"  [union-attr]
test/test_transaction.py:223: error: "Request" has no attribute "message"  [attr-defined]
test/test_transaction.py:225: error: Item "None" of "Any | None" has no attribute "addTransaction"  [union-attr]
test/test_transaction.py:226: error: Item "None" of "Any | None" has no attribute "delTransaction"  [union-attr]
test/test_transaction.py:226: error: "Request" has no attribute "transaction_id"  [attr-defined]
test/test_transaction.py:227: error: Item "None" of "Any | None" has no attribute "getTransaction"  [union-attr]
test/test_transaction.py:227: error: "Request" has no attribute "transaction_id"  [attr-defined]
test/test_transaction.py:243: error: Item "None" of "Any | None" has no attribute "processIncomingPacket"  [union-attr]
test/test_transaction.py:244: error: Item "None" of "Any | None" has no attribute "_buffer"  [union-attr]
test/test_transaction.py:258: error: Item "None" of "Any | None" has no attribute "processIncomingPacket"  [union-attr]
test/test_transaction.py:259: error: Item "None" of "Any | None" has no attribute "function_code"  [union-attr]
test/test_transaction.py:259: error: Item "None" of "Any | None" has no attribute "encode"  [union-attr]
test/test_transaction.py:273: error: Item "None" of "Any | None" has no attribute "processIncomingPacket"  [union-attr]
test/test_transaction.py:275: error: Item "None" of "Any | None" has no attribute "processIncomingPacket"  [union-attr]
test/test_transaction.py:291: error: Item "None" of "Any | None" has no attribute "processIncomingPacket"  [union-attr]
test/test_transaction.py:293: error: Item "None" of "Any | None" has no attribute "processIncomingPacket"  [union-attr]
test/test_transaction.py:309: error: Item "None" of "Any | None" has no attribute "processIncomingPacket"  [union-attr]
test/test_transaction.py:311: error: Item "None" of "Any | None" has no attribute "processIncomingPacket"  [union-attr]
test/test_transaction.py:328: error: Item "None" of "Any | None" has no attribute "processIncomingPacket"  [union-attr]
test/test_transaction.py:330: error: Item "None" of "Any | None" has no attribute "processIncomingPacket"  [union-attr]
test/test_transaction.py:349: error: Item "None" of "Any | None" has no attribute "processIncomingPacket"  [union-attr]
test/test_transaction.py:359: error: Cannot assign to a method  [method-assign]
test/test_transaction.py:366: error: Item "None" of "Any | None" has no attribute "buildPacket"  [union-attr]
test/test_transaction.py:368: error: Cannot assign to a method  [method-assign]
test/test_transaction.py:384: error: Item "None" of "Any | None" has no attribute "processIncomingPacket"  [union-attr]
test/test_transaction.py:386: error: Item "None" of "Any | None" has no attribute "processIncomingPacket"  [union-attr]
test/test_transaction.py:400: error: Item "None" of "Any | None" has no attribute "processIncomingPacket"  [union-attr]
test/test_transaction.py:414: error: Item "None" of "Any | None" has no attribute "processIncomingPacket"  [union-attr]
test/test_transaction.py:416: error: Item "None" of "Any | None" has no attribute "processIncomingPacket"  [union-attr]
test/test_transaction.py:430: error: Item "None" of "Any | None" has no attribute "processIncomingPacket"  [union-attr]
test/test_transaction.py:432: error: Item "None" of "Any | None" has no attribute "processIncomingPacket"  [union-attr]
test/test_transaction.py:439: error: Item "None" of "Any | None" has no attribute "decode_data"  [union-attr]
test/test_transaction.py:441: error: Item "None" of "Any | None" has no attribute "decode_data"  [union-attr]
test/test_transaction.py:457: error: Item "None" of "Any | None" has no attribute "processIncomingPacket"  [union-attr]
test/test_transaction.py:508: error: Item "None" of "Any | None" has no attribute "processIncomingPacket"  [union-attr]
test/test_transaction.py:514: error: Cannot assign to a method  [method-assign]
test/test_transaction.py:518: error: Item "None" of "Any | None" has no attribute "buildPacket"  [union-attr]
test/test_transaction.py:520: error: Cannot assign to a method  [method-assign]
test/test_transaction.py:536: error: Item "None" of "Any | None" has no attribute "processIncomingPacket"  [union-attr]
test/test_transaction.py:538: error: Item "None" of "Any | None" has no attribute "processIncomingPacket"  [union-attr]
test/test_transaction.py:552: error: Item "None" of "Any | None" has no attribute "processIncomingPacket"  [union-attr]
test/test_transaction.py:566: error: Item "None" of "Any | None" has no attribute "processIncomingPacket"  [union-attr]
test/test_transaction.py:568: error: Item "None" of "Any | None" has no attribute "processIncomingPacket"  [union-attr]
test/test_transaction.py:582: error: Item "None" of "Any | None" has no attribute "processIncomingPacket"  [union-attr]
test/test_transaction.py:583: error: Item "None" of "Any | None" has no attribute "_header"  [union-attr]
test/test_transaction.py:591: error: Cannot assign to a method  [method-assign]
test/test_transaction.py:596: error: Item "None" of "Any | None" has no attribute "buildPacket"  [union-attr]
test/test_transaction.py:598: error: Cannot assign to a method  [method-assign]
test/test_transaction.py:611: error: Item "None" of "Any | None" has no attribute "processIncomingPacket"  [union-attr]
test/test_transaction.py:625: error: Item "None" of "Any | None" has no attribute "processIncomingPacket"  [union-attr]
test/test_transaction.py:641: error: Item "None" of "Any | None" has no attribute "processIncomingPacket"  [union-attr]
test/test_transaction.py:658: error: Item "None" of "Any | None" has no attribute "processIncomingPacket"  [union-attr]
test/test_transaction.py:672: error: Item "None" of "Any | None" has no attribute "processIncomingPacket"  [union-attr]
test/test_transaction.py:686: error: Item "None" of "Any | None" has no attribute "processIncomingPacket"  [union-attr]
test/test_transaction.py:688: error: Item "None" of "Any | None" has no attribute "processIncomingPacket"  [union-attr]
test/test_transaction.py:694: error: Item "None" of "Any | None" has no attribute "populateResult"  [union-attr]
test/test_transaction.py:700: error: Cannot assign to a method  [method-assign]
test/test_transaction.py:705: error: Item "None" of "Any | None" has no attribute "buildPacket"  [union-attr]
test/test_transaction.py:707: error: Cannot assign to a method  [method-assign]
test/test_transaction.py:720: error: Item "None" of "Any | None" has no attribute "processIncomingPacket"  [union-attr]
Found 116 errors in 1 file (checked 1 source file)
```

**Review by janiversen — CHANGES_REQUESTED** [review]
We do not want type checking in test, as written to you before.

It does not bring any benefits to the production code, and you are sure to run into bigger problems, since we use advanced Pytest, which mypy by far do not support.

**Comment by janiversen** [comment]
Closing pull request as being not wanted. Thanks for your initiative, but it do help if you did read what 2 of use wrote to you before.

**Comment by jameshilliard** [comment]
> We do not want type checking in test, as written to you before.

I thought you didn't want type hinting in the tests, this is fixing actual bugs in the tests.

File: test/test_transaction.py [hunk 1]
```diff
@@ -22,27 +29,38 @@
...
 
-    client = None
-    decoder = None
-    _tcp = None
-    _tls = None
-    _rtu = None
-    _ascii = None
-    _manager = None
-    _tm = None
+    client: ModbusBaseSyncClient
```
Comment (alexrudd2) [inline test/test_transaction.py:32]: I don't see a need for type hinting the class attributes here at all; they will be defined in `setup_method()`.  Just removing the lines here satisfies `mypy`.

File: test/test_transaction.py [hunk 2]
```diff
@@ -22,27 +29,38 @@
...
+        self.client = mock.MagicMock(spec=ModbusBaseSyncClient)
+        self.client.framer = mock.MagicMock(spec=ModbusFramer)
         self.decoder = ServerDecoder()
-        self._tcp = ModbusSocketFramer(decoder=self.decoder, client=None)
-        self._tls = ModbusTlsFramer(decoder=self.decoder, client=None)
-        self._rtu = ModbusRtuFramer(decoder=self.decoder, client=None)
-        self._ascii = ModbusAsciiFramer(decoder=self.decoder, client=None)
-        self._manager = SyncModbusTransactionManager(self.client, 0.3, False, False, 3)
+        self._tcp = ModbusSocketFramer(
+            decoder=self.decoder, client=ModbusTcpClient("127.0.0.1")
```
Comment (alexrudd2) [inline test/test_transaction.py:49]: why initialize a client here?  I don't see how it improves anything over `client = None`

File: test/test_transaction.py [hunk 3]
```diff
@@ -114,56 +135,50 @@ def test_execute(self, mock_time):
         request.slave_id = 1
         request.function_code = 222
         trans = SyncModbusTransactionManager(client, 0.3, False, False, 3)
-        trans._recv = mock.MagicMock(  # pylint: disable=protected-access
-            return_value=b"abcdef"
-        )
+        setattr(trans, "_recv", mock.MagicMock(return_value=b"abcdef"))
```
Comment (alexrudd2) [inline test/test_transaction.py:138]: This is just tricking `mypy` and `pylint` which AFAIK don't process `setattr()`.

Cute, but not much advantage over `# type: ignore`.

File: test/test_transaction.py [hunk 4]
```diff
@@ -114,56 +135,50 @@ def test_execute(self, mock_time):
...
         assert trans.retries == 3
         assert not trans.retry_on_empty
 
-        trans.getTransaction = mock.MagicMock()
-        trans.getTransaction.return_value = "response"
+        setattr(trans, "getTransaction", mock.MagicMock(return_value=b"response"))
         response = trans.execute(request)
-        assert response == "response"
+        assert response == b"response"
```
Comment (alexrudd2) [inline test/test_transaction.py:144]: good catch!

File: test/test_transaction.py [hunk 5]
```diff
@@ -114,56 +135,50 @@ def test_execute(self, mock_time):
...
+        setattr(
+            trans, "_recv",
+            mock.MagicMock(side_effect=iter([b"abcdef", b"deadbe", b"123456"]))
         )
         client.comm_params.handle_local_echo = True
         trans.retry_on_empty = False
         trans.retry_on_invalid = False
-        assert trans.execute(request).message == "[Input/Output] Wrong local echo"
+        response = trans.execute(request)
+        assert isinstance(response, ModbusIOException)
```
Comment (alexrudd2) [inline test/test_transaction.py:167]: Initially (via email) I didn't like these `asserts`, but now I do think they are valid additions.

File: test/test_transaction.py [hunk 6]
```diff
@@ -195,33 +210,20 @@ def test_transaction_manager_tid(self):
...
         """Test the getting a transaction from the transaction manager."""
-
-        class Request:  # pylint: disable=too-few-public-methods
-            """Request."""
-
         self._manager.reset()
-        handle = Request()
-        handle.transaction_id = (  # pylint: disable=attribute-defined-outside-init
-            self._manager.getNextTID()
+        handle = ModbusRequest(
```
Comment (alexrudd2) [inline test/test_transaction.py:214]: This is a good improvement.

**Review by alexrudd2 — COMMENTED** [review]
I think some of the changes in here are good, and lead to a more robust test.  Some are just noise.

File: test/test_transaction.py (hunk 2 above)
Comment (jameshilliard) [reply test/test_transaction.py:49]: > why initialize a client here? I don't see how it improves anything over `client = None`

None isn't an allowable type for [`client: ModbusBaseSyncClient`](https://github.com/pymodbus-dev/pymodbus/blob/978ad88b2eb7482d681bbfcef46a8f0f05547f84/pymodbus/transaction.py#L137). Not sure if it's better to use a real client or a mock client here.

Comment (jameshilliard) [reply test/test_transaction.py:49]: I changed this to use `mock.create_autospec` for creating mock client objects.

Comment (alexrudd2) [reply test/test_transaction.py:49]: > > why initialize a client here? I don't see how it improves anything over `client = None`
> 
> `None` isn't an allowable type for [`client: ModbusBaseSyncClient`](https://github.com/pymodbus-dev/pymodbus/blob/978ad88b2eb7482d681bbfcef46a8f0f05547f84/pymodbus/transaction.py#L137). Not sure if it's better to use a real client or a mock client here.

Actually, only `_manager` needs a client (mock or real).  The others all accept a default argument of `client = None`

**Pull request closed without merging**